text/0106-artifact-indices.md

```diff
@@ -0,0 +1,175 @@
+
+It is crucial that we support the use-cases explained above, as the move to artifact bundles and processing in Symbolicator
+has introduced regressions.
+
+## Use existing Artifact Indexes
+
+Early iterations of the `artifact-lookup` API made use of artifact indexes. This was later removed as the API endpoint
+was not supposed to process large files. Downloading the artifact index to the API server, and performing lookups in
+it was deemed to cause problems.
+
+It was then replaced by just returning the _newest_ `N` release bundles. This has caused extreme load on the Database.
+After investigation, we found out that sorting these release bundles by date was two orders of magnitude slower than
+just returning `N` bundles unsorted. _Unsorted_ in this case would mean primary key order, which effective means
+returning the _oldest_ `N` bundles.
+
+Adding back support for artifact index lookups would temporarily solve the regression in SourceMap processing. However
+this is based on release bundles and artifact indices which are in the process of being phased out.
+
+- **pro**: Would allow querying / accessing the needed files.
+- **con**: Requires downloading / querying the index on API servers.
+- **con**: Based on deprecated technology that is in the process of being phased out.
+
+## Using a replacement flat-file index
+
+This would involve creating a replacement for artifact indexes based on the newer artifact bundles. Instead of a `JSON`
+index, a flat-file format was proposed. However the exact format is rather an insignificant detail.
+
+- **pro**: Would allow querying / accessing the needed files.
+- **pro**: Would solve one pain-point of artifact indexes: No more `Release` object needed.
+- **con**: Requires downloading / querying the index on API servers.
+- **con**: Needs to re-create / replicate most of the artifact-index logic, and inherits all of its cons:
+  - updating / merging the index on uploads
+  - the index growing monotonically on disjoint uploads
+
+## Index individual files in the Database
+
+This would create a similar database table / index to `ReleaseFile`, but with different constraints.
+
+- **pro**: Files individually indexed.
+- **pro**: A database-based solution might be simpler than maintaining a separate index manually.
+- **pro**: No `Release` object needed.
+- **pro**: No need to download / process any index on API servers.
+- **pro**: Would work well with auto-expiration / partitioning if implemented correctly.
+- **con**: High load on Database and its indices.
+- **con**: Might require a custom solution to better compress `release` and `url`, possibly by using some kind of interning.
+
+The table could look something the following:
+
+- org / project
+- (optional) release + dist
+- url
+- date_added / touched
+- artifact bundle id
+
+## Hybrid Bundles / Individual files
+
+The idea was brought up to only index individual files if the number of uploaded bundles per release exceeds a certain
+threshold `N`.
```

Comment:
```suggestion
The idea was brought up to only index individual files if the number of uploaded bundles per release exceeds a certain
threshold `N` (which could easily be `1` since we want to optimize the case in which there is no need to create an index, since we can just leverage the `manifest.json` in the bundle).
```